fix(compiler): non-literal inline templates incorrectly processed in partial compilation

devversion opened the pull request:
Currently if a component defines a template inline, but not through a
string literal, the partial compilation references the template expression
as is. This is problematic because the component declaration can no longer
be processed by the linker later as there is no static interpretation. e.g.

```js
const myTemplate = `...`;

TestCmp.ɵcmp = i0.ɵɵngDeclareComponent({
  version: "0.0.0-PLACEHOLDER",
  type: TestCmp,
  selector: "test-cmp",
  ngImport: i0,
  template: myTemplate,
  isInline: true
});
```

To fix this, we use the the resolved template in such cases so that
the linker can process the template/component declaration as expected.

Note: I considered always passing the `templateStr` as part of `ParsedTemplate` while
also passing a `sourceSpan`. This looked like the most clean solution to me initially, but
turned out to be not suitable because it would mean that we always create a new `o.literal`.
This literal can be different to original inline template literal's (i.e. due to quotes) and therefore
causing incorrect source mappings for the template.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -1432,4 +1415,4 @@ function makeCyclicImportInfo(
       `The ${type} '${name}' is used in the template but importing it would create a cycle: `;
   return makeRelatedInformation(ref.node, message + path);
-}
+}
```
Where did the newline go? :-)

devversion replied:
haha. it's back now 😄

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/content_projection/TEST_CASES.json:
```diff
@@ -124,6 +124,34 @@
           ]
         }
       ]
+    },
```
I think these tests fit better directly under `components_and_directives` instead of within `content_project`, so perhaps these tests can be moved one directory up?

devversion replied:
yeah. I put them accidentally inside there and didn't notice

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/content_projection/TEST_CASES.json:
```diff
@@ -124,6 +124,34 @@
       ]
+    },
+    {
+      "description": "should support inline non-literal templates (partial compile)",
```
I don't see a difference in the expectation file between the partial and full compile, so I think these tests can be merged and have no `compilationModeFilter`?

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/content_projection/non_literal_template.ts:
```diff
@@ -0,0 +1,7 @@
+import {Component} from '@angular/core';
+
+const myTemplate = `<div *ngIf="show">Hello</div>`;
```
Can you also add some more variants:

- Substitution template literal: `<div *ngIf="show">${'Hello'}</div>`;
- Dynamic template: `` '<div *ngIf="show">' + 'Hello' + '</div>';``;

And perhaps also the above two but inlined into the `template` property's expression.

As an aside; the template literal you currently have is actually quite interesting because that one could allow for accurate source mapping. We don't currently support it, but we could if we wanted to 😄  (it's non-trivial though as we'd need to keep track of AST nodes in the static interpreter)

devversion replied:
yeah, I figured there would be a way to get access to the referenced literal 😄 Maybe a future project for angular

JoostK reviewed (commented):
Awesome, thank you so much @devversion!

petebacondarwin commented:
As discussed offline - @devversion is going to refactor how the information is passed between the compiler-cli component annotation and the compiler (e.g. for `compileDeclareComponentFromMetadata()` to try to avoid giving `parseTemplate()` too much knowledge/responsibility.

devversion committed:
fix(compiler): non-literal inline templates incorrectly processed in partial compilation

Currently if a component defines a template inline, but not through a
string literal, the partial compilation references the template expression
as is. This is problematic because the component declaration can no longer
be processed by the linker later as there is no static interpretation. e.g.

```js
const myTemplate = `...`;

TestCmp.ɵcmp = i0.ɵɵngDeclareComponent({
  version: "0.0.0-PLACEHOLDER",
  type: TestCmp,
  selector: "test-cmp",
  ngImport: i0,
  template: myTemplate,
  isInline: true
});
```

To fix this, we use the the resolved template in such cases so that
the linker can process the template/component declaration as expected.

devversion committed:
refactor(compiler): clean up template information passed for partial compilation

With the introduction of the partial compilation, the Angular compiler's
existing `parseTemplate` method has been extended to pass through multiple
properties purely in favor of the partial compilation.

e.g. the `parseTemplate` function now accepts an "option" called `isInline`.
This option is just passed through and returned as part of the `ParsedTemplate`.

This is not ideal because the `parseTemplate` function doesn't care
whether the specified template was inline or not. This commit cleans
up the `parseTemplate` compiler function so that nothing needed only
for the partial compilation is added to it.

We introduce a new struct for additional template information that
is specific to the generation of the `declareComponent` function. With
that change, we can simplify the component decorator handler and keep
logic more local.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -1164,13 +1164,21 @@ export class ComponentDecoratorHandler implements
+    // If the template has been defined inline and through a string literal, we capture it
+    // in the parsed component template. This is useful for partial compilation which can
+    // re-use the original literal for proper template source mapping.
+    const inlineTemplateLiteralExpression =
+        template.isInline && ts.isStringLiteralLike(template.expression) ?
+        new WrappedNodeExpr(template.expression) :
+        null;
```
Could we rename this to `templateExpression` and define it via:

```ts
const templateExpression =
        template.isInline && (
          ts.isStringLiteralLike(template.expression) ?
            new WrappedNodeExpr(template.expression) :
            o.literal(templateInfo.content, null, null)
          ) : null;
```

Then we don't need to do it in the partial compiler.

devversion replied:
As per offline discussion: We'd want to keep the logic local to the partial compilation code, but we can even make things more local by composing the `DeclareComponentTemplateInfo` object directly in `compilePartial`.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -1053,21 +1062,19 @@ export class ComponentDecoratorHandler implements
-      let templateLiteral: ts.Node|null = null;
-      let templateUrl: string = '';
-      let templateRange: LexerRange|null = null;
+      let sourceStr: string;
+      let sourceParseRange: LexerRange|null = null;
+      let templateContent: string;
```
Any reason for these renames? They don't seem to make much difference and just add churn to the commit diff.

devversion replied:
I've renamed them so that it's easier to figure out what each variable means. A few of these variables were unused, and with the refactorings, I want to make it clear that `templateContent` is different to `templateStr`. Those are two separate things if a lexer range is specified. With the new names I think it's pretty clear what is meant.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -1079,38 +1086,45 @@ export class ComponentDecoratorHandler implements
         this.depTracker.addResourceDependency(
             node.getSourceFile(), absoluteFrom(template.resolvedTemplateUrl));
       }
+      // For external template files, the template content is equal to the source content.
+      const templateContent = sourceStr;
```
This seems like a superfluous additional variable?

devversion replied:
I've added it just for more readability.. though I realize it's not strictly needed. updating

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -1150,26 +1163,22 @@ export class ComponentDecoratorHandler implements
     return {
       ...parsedTemplate,
       diagNodes,
-      template: template.isInline ? new WrappedNodeExpr(template.expression) : templateStr,
-      templateUrl: template.resolvedTemplateUrl,
-      isInline: template.isInline,
-      file: new ParseSourceFile(templateStr, template.resolvedTemplateUrl),
```
🎉

petebacondarwin commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -2438,4 +2393,14 @@ export interface ParsedTemplate {
    */
   commentNodes?: t.Comment[];
-}
+
+  /**
+   * Whether whitespace nodes are included in the parsed output.
+   */
+  preserveWhitespaces?: boolean;
+
+  /**
+   * Configuration that has been used to parse interpolations in the template.
+   */
+  interpolationConfig?: InterpolationConfig;
+}
```
These appear to have just been moved from the top of the interface. Was that intentional?

devversion replied:
just a result of the many iterations I've undergone.

petebacondarwin commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -2082,6 +2082,7 @@ export interface ParseTemplateOptions {
    * `$localize` message id format and you are not using compile time translation merging.
    */
   enableI18nLegacyMessageIdFormat?: boolean;
+
```
❤️

petebacondarwin reviewed (changes requested):
Thanks @devversion - this goes a long way to tidying up this area of the compiler and fixes the issue. I have left a few suggestions.

devversion commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/non_literal_template_with_concatenation.ts:
```diff
@@ -0,0 +1,8 @@
+import {Component} from '@angular/core';
+
+const greeting = 'Hello!';
+const myTemplate = `<div *ngIf="show">${greeting}</div>`;
```
oops. yeah! good catch

devversion commented on packages/compiler/src/render3/partial/component.ts:
```diff
@@ -42,8 +42,11 @@ export interface DeclareComponentTemplateInfo {
-  /** Expression that resolves to the inline template. */
-  inlineTemplateExpression: o.Expression|null;
+  /**
+   * Expression that defined the component template inline as a literal. Null if
+   * the template has not been defined inline through a literal.
```
SGTM

devversion commented on packages/compiler/src/render3/partial/component.ts:
```diff
@@ -111,19 +114,29 @@ export function createComponentDefinitionMap(
 
 function getTemplateExpression(
     template: ParsedTemplate, templateInfo: DeclareComponentTemplateInfo): o.Expression {
+  // If the template has been defined using a literal, we use that expression directly
```
👍

devversion commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -7,6 +7,7 @@
  */
 
 import {compileClassMetadata, compileComponentFromMetadata, compileDeclareClassMetadata, compileDeclareComponentFromMetadata, ConstantPool, CssSelector, DeclarationListEmitMode, DEFAULT_INTERPOLATION_CONFIG, DomElementSchemaRegistry, Expression, ExternalExpr, FactoryTarget, InterpolationConfig, LexerRange, makeBindingParser, ParsedTemplate, ParseSourceFile, parseTemplate, R3ClassMetadata, R3Compo ...
+import {DeclareComponentTemplateInfo} from '@angular/compiler/src/compiler';
```
it is, I just relied on VScode to auto-import but didn't check.

petebacondarwin commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -2438,4 +2402,4 @@ export interface ParsedTemplate {
    */
   commentNodes?: t.Comment[];
-}
+}
```
😱

devversion replied:
😱 My excuse is that I'm using a new machine since my windows machine died!

petebacondarwin reviewed (approved):
LGTM! Nice work @devversion - thanks for your patience.

devversion committed:
fixup! refactor(compiler): clean up template information passed for partial compilation

Add missing new line

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/non_literal_template_with_substitution.js:
```diff
@@ -0,0 +1,7 @@
+function TestCmp_div_0_Template(rf, ctx) {
+  if (rf & 1) {
+    i0.ɵɵelementStart(0, "div");
+    i0.ɵɵtext(1, "Hello!");
```
It is indeed:

https://github.com/angular/angular/blob/c7c5b2fc1e1cf4960d897e2b0432769deaac96a3/packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts#L202-L215

JoostK reviewed (approved):
Nice cleanup and thanks for the fix!

AndrewKushnir commented:
[Presubmit](http://test/OCL:368748914:BASE:368748923:1618532227654:95cf23e7).

AndrewKushnir commented:
@devversion FYI presubmits are successful for the changes in this PR.
